blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeGraphVisitor.java

```diff
@@ -323,20 +337,44 @@ public void parallelStart(@Nonnull FlowNode parallelStartNode, @Nonnull FlowNode
             }
 
             branch.setPipelineActions(branchActions);
-
-            if(nextStage!=null) {
+            // do we have nested sequential stages for this parallel branch?
+            Stack<FlowNodeWrapper> stack = stackPerEnd.get(endNode.getId());
+            if(stack!=null&&!stack.isEmpty()){
+                // yes so we can rebuild the graph here
+                // we don't want the first one as it's a duplicate of the parallel parent but with stage type so rid of it
+                FlowNodeWrapper flowNodeWrapper = stack.pop();
+                if(stack.isEmpty()){
+                    if(nextStage!=null) {
+                        branch.addEdge(nextStage);
+                    }
+                    parallelBranches.push(branch);
+                    continue;
+                }
+                // if name is different it's not a dummy 'duplicate' stage with the same name so let's add it to the graph
+                if(!StringUtils.equals( flowNodeWrapper.getDisplayName(), branch.getDisplayName())){
```

Comment:
Probably should use `flowNodeWrapper.getId()` instead of `flowNodeWrapper.getDisplayName()`?

Reply:
I wish but not really :(
In such case:
```
pipeline {
    agent any
    stages {
        stage("top") {
            parallel {
                stage("first") {
                   ......
                }
    .......
```
We have 2 nodes:
* StepStartNode (descriptorId: org.jenkinsci.plugins.workflow.cps.steps.ParallelStep) and ParallelLabelAction "first"
* StepStartNode (descriptorId: org.jenkinsci.plugins.workflow.support.steps.StageStep) and LabelAction "first"
id are very different. Anyway the test should happen only in this special case

Reply:
I was suggesting using ‘flowNode.getId()’ which is always unique for each FlowNode, I guess you are saying that’s not the case?

Reply:
yup. but the need is to eliminate the extra node because we only want one node with type parallel ("first") and not the one with type stage.